rpc/attestation/middleware.go

```diff
@@ -18,29 +19,41 @@ import (
 // If the HTTP request includes an X-Attestation-Nonce header, its value is sent to the NSM and included in
 // the final attestation document.
 func Middleware(enc *enclave.Enclave) func(http.Handler) http.Handler {
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			ctx := r.Context()
+	runMiddleware := func(w http.ResponseWriter, r *http.Request) (ctx context.Context, cancelFunc func() error, err error) {
+		ctx, span := tracing.Span(r.Context(), "attestation.Middleware")
+		defer func() {
+			if err != nil {
+				span.RecordError(err)
+			}
+			span.End()
+		}()
 
-			var nonce []byte
-			if nonceVal := r.Header.Get("X-Attestation-Nonce"); nonceVal != "" {
-				nonceVal = strings.TrimSpace(nonceVal)
-				if len(nonceVal) > 32 {
-					proto.RespondWithError(w, fmt.Errorf("X-Attestation-Nonce value cannot be longer than 32"))
-					return
-				}
-				nonce = []byte(nonceVal)
+		var nonce []byte
+		if nonceVal := r.Header.Get("X-Attestation-Nonce"); nonceVal != "" {
+			nonceVal = strings.TrimSpace(nonceVal)
+			if len(nonceVal) > 32 {
+				return nil, nil, fmt.Errorf("X-Attestation-Nonce value cannot be longer than 32")
 			}
+			nonce = []byte(nonceVal)
+		}
+
+		att, err := enc.GetAttestation(ctx, nonce)
+		if err != nil {
+			return nil, nil, err
+		}
 
-			att, err := enc.GetAttestation(ctx, nonce)
+		w.Header().Set("X-Attestation-Document", base64.StdEncoding.EncodeToString(att.Document()))
+		return context.WithValue(r.Context(), contextKey, att), att.Close, nil
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx, cancel, err := runMiddleware(w, r)
```

Comment:
Yes, we want the span to be created only for the middleware and so that it ends before calling the next handler. Otherwise every handler's span would be a child of the previous one